tiledb/sm/cpp_api/ndrectangle.h

```diff
@@ -0,0 +1,229 @@
+/**
+ * @file   ndrectangle.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the C++ API for the TileDB NDRectangle.
+ */
+
+#ifndef TILEDB_CPP_API_NDRECTANGLE_H
+#define TILEDB_CPP_API_NDRECTANGLE_H
+
+#include "context.h"
+#include "deleter.h"
+#include "dimension.h"
+#include "domain.h"
+#include "tiledb.h"
+#include "type.h"
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace tiledb {
+
+class NDRectangle {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /**
+   * Constructor
+   *
+   * @param ctx The TileDB context
+   * @param domain The input Domain
+   */
+  explicit NDRectangle(const tiledb::Context& ctx, const tiledb::Domain& domain)
+      : ctx_(ctx) {
+    tiledb_ndrectangle_t* capi_ndrect;
+    ctx.handle_error(tiledb_ndrectangle_alloc(
+        ctx.ptr().get(), domain.ptr().get(), &capi_ndrect));
+    ndrect_ = std::shared_ptr<tiledb_ndrectangle_t>(capi_ndrect, deleter_);
+  }
+
+  /**
+   * Constructor
+   *
+   * @param ctx The TileDB context
+   * @param ndrect The NDRectangle C API pointer
+   *
+   * @note The object will manage the lifetime of tiledb_ndrectangle_t*
+   */
+  NDRectangle(const tiledb::Context& ctx, tiledb_ndrectangle_t* ndrect)
+      : ctx_(ctx) {
+    ndrect_ = std::shared_ptr<tiledb_ndrectangle_t>(ndrect, deleter_);
+  }
+
+  NDRectangle(const NDRectangle&) = default;
+  NDRectangle(NDRectangle&&) = default;
+  NDRectangle& operator=(const NDRectangle&) = default;
+  NDRectangle& operator=(NDRectangle&&) = default;
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /**
```

Comment:
NIT: the line breaks here are weird... Let's try to align closer to the 80 character limit and see if we can fit on 2 lines.